src/ucs/arch/generic/cpu.c

```diff
@@ -0,0 +1,109 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/arch/cpu.h>
+#include <ucs/arch/generic/cpu.h>
+#include <ucs/sys/sys.h>
+#include <ucs/sys/string.h>
+#include <ucs/sys/stubs.h>
+#include <ucs/type/init_once.h>
+
+#define UCS_CPU_CACHE_FILE_FMT     "/sys/devices/system/cpu/cpu%d/cache/index%d/%s"
+#define UCS_CPU_CACHE_LEVEL        "level"
+#define UCS_CPU_CACHE_TYPE         "type"
+#define UCS_CPU_CACHE_SIZE         "size"
+
+static ucs_init_once_t ucs_cache_read_once = UCS_INIT_ONCE_INIITIALIZER;
+
+/* cache datatypes */
+struct { /* cache info obtained from [linux] kernel */
+    int             level;
+    const char     *type;
+} const ucs_cpu_cache_kernel[] = {
+    [UCS_CPU_CACHE_L1d] = {.level = 1, .type = "Data"},
+    [UCS_CPU_CACHE_L1i] = {.level = 1, .type = "Instruction"},
+    [UCS_CPU_CACHE_L2]  = {.level = 2, .type = "Unified"},
+    [UCS_CPU_CACHE_L3]  = {.level = 3, .type = "Unified"}
+};
+
+ucs_cache_size_t ucs_cpu_cache_size[] = { /* cached values of CPU cache */
+    [UCS_CPU_CACHE_L1d] = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L1i] = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L2]  = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED},
+    [UCS_CPU_CACHE_L3]  = {.cache_size = 0, .status = UCS_ERR_UNSUPPORTED}
+};
+
+
+static void ucs_cpu_read_cache_size(ucs_cache_size_t *cache, size_t length)
+{
+    char type_str[32];  /* Data/Instruction/Unified */
+    char size_str[32];  /* memunits */
+    int i;
+    int cpu;
+    long level;
+    ssize_t file_size;
+    size_t c;
+    ucs_status_t status;
+
+    /* try first CPU-specific algorithm */
+    ucs_arch_get_cache_size(cache, length);
```

Comment:
usually specific arch calls generic, not the other way around

Reply:
ok, file is moved into arch directory